### PR TITLE
synccontroller: trigger the delete operation when the relate resource does not exist

### PR DESCRIPTION
**Pull request by zc2638**
Signed-off-by: zc <ce.zheng@daocloud.io>

**What type of PR is this?**

/kind feature

**What this PR does / why we need it**:
 When the edge side communication is disconnected and the pod is restarted, edge cannot receive the request and confirm. The historical pod is deleted and a new pod is generated. When the edge is ready, the pod will be distributed and a new object resource will be generated. The historical resource redundancy is only judged to skip and should be deleted.
 Edgecontroller manages object resources. It should handle invalid object resources.

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note

```

### CODE REVIEW

**Review by fisherxu — COMMENTED** [review]
Thanks for the contributions, and could you please add more descriptions for the PR?

**Comment by zc2638** [comment]
@fisherxu @Rachel-Shao yeah, this PR is dealing with the problem

**Comment by GsssC** [comment]
Overall lgtm
and When the object will be delete if the relate resource does not exit?

**Comment by zc2638** [comment]
> Overall lgtm
> and When the object will be delete if the relate resource does not exit?

Synccontroller synchronizes `object` resources every 5 seconds.

```golang
func (sctl *SyncController) Start() {
	if !cache.WaitForCacheSync(beehiveContext.Done(), sctl.informersSyncedFuncs...) {
		klog.Errorf("unable to sync caches for sync controller")
		return
	}

	sctl.deleteObjectSyncs() //check outdate sync before start to reconcile
	go wait.Until(sctl.reconcile, 5*time.Second, beehiveContext.Done())
}
```

**Comment by GsssC** [comment]
Sorry I made a mistake， When the **sync object** will be delete if the relate resource does not exsit?

**Comment by zc2638** [comment]
/cc @fisherxu 
need reopen

File: cloud/pkg/synccontroller/objectsync.go [hunk 1]
```diff
@@ -34,46 +36,59 @@ func (sctl *SyncController) manageObject(sync *v1alpha1.ObjectSync) {
...
 		return
 	}
-
 	syncObjUID := getObjectUID(sync.Name)
 	if syncObjUID != string(object.GetUID()) {
 		err = apierrors.NewNotFound(schema.GroupResource{
 			Group:    "",
 			Resource: sync.Spec.ObjectKind,
 		}, sync.Spec.ObjectName)
+		sendFailEvents(sync, err)
```
Comment (chenchunxiu) [inline cloud/pkg/synccontroller/objectsync.go:55]: just question：why sendFailEvent here to send delete msg to edge-node? metaserver will delete from meta db according to msg.key not obj.uid.

File: cloud/pkg/synccontroller/objectsync.go [hunk 2]
```diff
@@ -34,46 +36,59 @@ func (sctl *SyncController) manageObject(sync *v1alpha1.ObjectSync) {
...
+		msg := buildEdgeControllerMessage(nodeName, sync.Namespace, sync.Spec.ObjectKind, sync.Spec.ObjectName, model.DeleteOperation, obj)
 		beehiveContext.Send(commonconst.DefaultContextSendModuleName, *msg)
-		return
 	}
+	// TODO Other error handling methods
+}
 
+func sendEvents(sync *v1alpha1.ObjectSync, objectResourceVersion string, obj interface{}) {
+	nodeName := getNodeName(sync.Name)
+	resourceType := sync.Spec.ObjectKind
```
Comment (chenchunxiu) [inline cloud/pkg/synccontroller/objectsync.go:83]: just question: shoud we do something about `sync.Spec.ObjectKind`?  
E.g `sync.Spec.ObjectKind` may be “Pod”， I think here need “pod” or “pods”.

**Comment by chenchunxiu** [comment]
@zc2638 could you reopen this pr?

File: cloud/pkg/synccontroller/objectsync.go (hunk 2 above)
Comment (zc2638) [reply cloud/pkg/synccontroller/objectsync.go:83]: Yes, the `metamanager` will handle it. The database `meta` table stores kind.

File: cloud/pkg/synccontroller/objectsync.go (hunk 1 above)
Comment (zc2638) [reply cloud/pkg/synccontroller/objectsync.go:55]: This is managed by `metamanager`

File: cloud/pkg/synccontroller/objectsync.go [hunk 3]
```diff
@@ -36,58 +36,59 @@ func (sctl *SyncController) manageObject(sync *v1alpha1.ObjectSync) {
...
-			obj.SetNamespace(sync.Namespace)
-		} else {
-			klog.Errorf("Failed to manage pod sync of %s in namespace %s: %v", sync.Name, sync.Namespace, err)
-			return
-		}
-	}
-	sendEvents(err, nodeName, sync, sync.Spec.ObjectKind, object.GetResourceVersion(), object)
+		//trigger the delete event
+		klog.Infof("%s: %s has been deleted in K8s, send the delete event to edge", sync.Spec.ObjectKind, sync.Spec.ObjectName)
+		msg := buildEdgeControllerMessage(nodeName, sync.Namespace, sync.Spec.ObjectKind, sync.Spec.ObjectName, model.DeleteOperation, obj)
```
Comment (neiba) [inline cloud/pkg/synccontroller/objectsync.go:75]: cloudcore would crash when msg is nil, if you pod in edge use a pvc, and i found pvc's objectsync Spec.ObjectKind is empty, so buildEdgeControllerMessage return nil.
here, you should handle condition if msg is nil

**Pull request closed without merging**